examples/cosmos/ics23/hashes.qnt

```diff
@@ -0,0 +1,156 @@
+/// A symbolic specification of a string that is suitable for reasoning about
+/// strings and hashes (such as SHA256). The exact hash function is irrelevant.
+/// The only assumption is that a hash is irreversible.
+///
+/// Similar to symbolic reasoning by Dolev and Yao, we would like to represent:
+///  - a raw sequence of bytes, e.g., [ 1, 2, 3 ],
+///  - a hashed sequence, e.g., h([ 1, 2, 3 ]),
+///  - a concatenation of a raw sequence and a hash (in any order),
+///    e.g., [ 1, 2 ] + h([ 3, 4 ])],
+///  - nested hashes, e.g., h([ 1, 2 ] + h([ 3, 4 ])).
+///
+/// So the idea is basically to represent words as terms, where the atoms are
+/// sequences of integers, and terms are constructed as n-ary applications
+/// of the symbol "h", which means "hash". Since Quint enforces strict typing
+/// and it does not allow for inductive/recursive algebraic data types,
+/// we represent terms as maps that encode trees. For instance,
+/// the term h([ 1, 2 ] + h([ 3, 4 ])) + [ 5, 6 ] corresponds to the tree:
+///
+///            *
+///          /   \
+///       "h"     [ 5, 6 ]
+///     /     \
+/// [ 1, 2 ]   "h"
+///          /
+///      [ 3, 4 ]
+///
+/// The above tree is represented as a map in Quint. Each key corresponds to a path in the tree; for example, the first root's child is [ 0 ], the second root's child is [ 1 ], the first child of [ 0 ] is [ 0, 0], etc.
+///
+/// Map([ 0 ] -> { t: "h", b: [] },
+///     [ 0, 0 ] -> { t: "r", b: [ 1, 2 ] },
+///     [ 0, 1 ] -> { t: "h", b: [] } ),
+///     [ 0, 1, 0 ] -> { t: "r", b: [ 3, 4 ] },
+///     [ 1 ] -> { t: "r", b: [ 5, 6 ] })
+///
+/// Igor Konnov, Informal Systems, 2022-2023
+module hashes {
+  /// A sequence of bytes is simply a list of integers
+  type Bytes_t = List[int]
+
+  /// compare two lists of integers (e.g., bytes) lexicographically
+  pure def lessThan(w1: Bytes_t, w2: Bytes_t): bool = {
+    pure val len1 = length(w1)
+    pure val len2 = length(w2)
+    or {
+      len1 < len2 and indices(w1).forall(i => w1[i] <= w2[i]),
+      and {
+        len1 == len2,
+        indices(w1).exists(i => and {
+          w1[i] < w2[i],
+          indices(w1).forall(j => j < i implies w1[j] == w2[j])
+        })
+      }
+    }
+  }
+
+  /// Returns true iff `w` is the nil value.
+  /// In our spec, nil is equivalent to the empty list.
+  pure def isNil(w: Bytes_t): bool = {
+    w == []
+  }
+
+  /// A tree node that represents a fragment of a term
+  type TERM_NODE_T = {
+    // node type: either "h" (for hash), or "r" (for raw sequence)
+    t: str,
+    // the bytes when t = "r", and [] when t = "h"
+    b: Bytes_t
+  }
+
+  /// A word is a map from a path to the term node.
+  /// The first root's child is [ 0 ], the second root's child is [ 1 ],
+  /// the first child of [ 0 ] is [ 0, 0], etc.
+  type Term_t = Bytes_t -> TERM_NODE_T
```

Comment:
I have to note that Quint sum types are not inductive/recursive, for a reason of not admitting unbounded data structures. So it would not be possible to encode a tree with Quint sum types.

Reply:
But now it is a list of integers, where only 0 and 1 is used. Why cannot it be a list of a sum type instead?

Reply:
Try to write it in any language that has complete support for sum types, e.g., in OCaml. You will see that you need recursion. Basically, you have to say that a node refers to the children nodes (of the same type). If you use a mathematical function instead, e.g., $f: Paths \rightarrow Nodes$, you do not need an inductive definition.

Reply:
A list of a sum type would not give you much, as you would not want a leaf to appear in the middle of the list.